Doc(pyavd): Add porting guide for breaking pyavd changes (#6442)
aristanetworks/ansible-avd: docs/porting-guides/6.x.x.md

## Changes to PyAVD

> PyAVD has undergone a number of changes in AVD 6.0.0. Only the changes requiring action are listed here.

Reviewer: Any updates needed for `docs/_media/pyavd_functions_dark.svg` & `docs/_media/pyavd_functions_light.svg` to reflect that "validated inputs" returned by the `validate_inputs()` are expected to be passed into `get_avd_facts()` and `get_device_structured_config` instead of passing raw inputs?

Author: good point I should update that.

Author: please take a look at the updated drawings. @JulioPDX I see some font changes I did not do. I prefer the monospaced one for these function names.

Reviewer: Looks great to me!

### Changes to `pyavd.validate_inputs` and `pyavd.validate_structured_config`

In AVD 6.0.0 the validation functions now return the validated and coerced data instead of updating it in-place. The returned object is now a `ValidatedDataResult`.
This means the caller must use the returned data instead of the original data for subsequent calls to PyAVD functions.

See the PyAVD [migration example](#example-of-migration-from-pyavd-5x-to-600-for-a-small-build-script) for more information.

### Changes to `pyavd.get_avd_facts`

> In AVD 6.0.0, `pyavd.get_avd_facts` now accepts either a dictionary of `AVDDesign` dataclasses or a dictionary of dictionaries.

Reviewer 2: I don't think `AVDDesign` and `EOSConfig` are Python dataclasses, are they?
```suggestion
In AVD 6.0.0, `pyavd.get_avd_facts` now accepts either a dictionary of `AVDDesign` instances or a dictionary of dictionaries.
```

When using custom Python modules for IP Addressing or Interface Descriptions, the `all_hostvars` must also be passed to `pyavd.get_avd_facts` so the Python modules can access the hostvars.

See the PyAVD [migration example](#example-of-migration-from-pyavd-5x-to-600-for-a-small-build-script) for more information.

### Changes to `pyavd.get_device_structured_config`

> In AVD 6.0.0, `pyavd.get_device_structured_config` now returns an `EOSConfig` dataclass instead of a dictionary.

Reviewer 2: ```suggestion
In AVD 6.0.0, `pyavd.get_device_structured_config` now returns an `EOSConfig` instance instead of a dictionary.
```

> This dataclass holds the complete structured configuration for a device and can be used directly as input

Reviewer 2: ```suggestion
This instance holds the complete modeled structured configuration for a device and can be used directly as input
```

- serialized_settings = settings.model_dump()
+ serialized_settings = asdict(settings)
```

### Example of migration from PyAVD 5.x to 6.0.0 for a small build script

The tabs below shows a a simple example script to build a small fabric from generated inputs.

The script is first shown for PyAVD 5.x, then for PyAVD 6.0.0 and finally the diff to migrate from 5.x to 6.0.0.

> !!! note

Reviewer 2: ```suggestion
!!! warning
```

    The example script is not meant to be a complete or optimal build script
    but rather to show the changes required to migrate from PyAVD 5.x to 6.0.0.

=== "PyAVD v5.x"
    --8<-- docs/pyavd/simple_example_5.x.md

=== "PyAVD v6.0.0"
    --8<-- docs/pyavd/simple_example_6.x.md

=== "PyAVD v5.x to 6.0.0"

>     --8<-- docs/pyavd/simple_example_5_to_6.md

Reviewer 2: There is a typo in your file name: `simple_example_5.x.md` versus `simple_example_5_x.md` notice the `_x` instead of `.x`. You also need opening and closing markers.
```suggestion
=== "PyAVD v5.x"

    --8<--
    docs/pyavd/simple_example_5.x.md
    --8<--

=== "PyAVD v6.0.0"

    --8<--
    docs/pyavd/simple_example_6.x.md
    --8<--

=== "PyAVD v5.x to 6.0.0"

    --8<--
    docs/pyavd/simple_example_5_to_6.md
    --8<--
```

Author: single line format is accepted by the plugin.
https://facelessuser.github.io/pymdown-extensions/extensions/snippets/#single-line-format

- serialized_settings = settings.model_dump()
+ serialized_settings = asdict(settings)
```

### Example of migration from PyAVD 5.x to 6.0.0 for a small build script

> The tabs below shows a a simple example script to build a small fabric from generated inputs.

Reviewer: ```suggestion
The tabs below show a simple example script to build a small fabric from generated inputs.
```
